File: tests/bench/WordFreq.java
// Same program as word_freq.jai.
import java.util.HashMap;
import java.util.Map;

public class WordFreq {
    public static void main(String[] args) {
        StringBuilder builder = new StringBuilder();
        long seed = 7;
        for (int i = 0; i < 200000; i++) {
            seed = (seed * 1103515245L + 12345L) % 2147483648L;
            builder.append("w").append(seed % 500L).append(" ");
        }
        String text = builder.toString();

        Map<String, Long> counts = new HashMap<>();
        int start = 0;
        int at = 0;
        int n = text.length();
        while (at < n) {
            if (text.charAt(at) == ' ') {
                if (at > start) {
                    String word = text.substring(start, at);
                    counts.put(word, counts.getOrDefault(word, 0L) + 1L);
                }
                start = at + 1;
            }
            at++;
        }

        long total = 0;
        for (Map.Entry<String, Long> entry : counts.entrySet()) {
            total += entry.getValue();
        }
        System.out.println(counts.size());
        System.out.println(total);
    }
}

File: tests/bench/word_freq.cpp
// Same program as word_freq.jai. See tests/bench/README.md for why the C++ and
// Java rows are here and what they are not.
#include <cstdint>
#include <cstdio>
#include <string>
#include <unordered_map>

int main() {
    std::string text;
    int64_t seed = 7;
    for (int64_t i = 0; i < 200000; i++) {
        seed = (seed * 1103515245 + 12345) % 2147483648;
        char buf[32];
        std::snprintf(buf, sizeof buf, "w%lld ", (long long)(seed % 500));
        text += buf;
    }

    std::unordered_map<std::string, int64_t> counts;
    size_t start = 0;
    size_t at = 0;
    size_t n = text.size();
    while (at < n) {
        if (text[at] == ' ') {
            if (at > start) {
                std::string word = text.substr(start, at - start);
                auto found = counts.find(word);
                int64_t previous = found == counts.end() ? 0 : found->second;
                counts[word] = previous + 1;
            }
            start = at + 1;
        }
        at++;
    }

    int64_t total = 0;
    for (const auto &entry : counts) total += entry.second;
    std::printf("%lld\n", (long long)counts.size());
    std::printf("%lld\n", (long long)total);
    return 0;
}
